doc/source/reference/c-api.types-and-structures.rst

```diff
@@ -377,13 +380,28 @@ PyArrayDescr_Type
     normally a Python string. These tuples are placed in this
     dictionary keyed by name (and also title if given).
 
+.. c:member:: PyObject *PyArray_Descr.names
+
+    A table of functions specifiv for each
+    basic data descriptor.
+
 .. c:member:: PyArray_ArrFuncs *PyArray_Descr.f
 
     A pointer to a structure containing functions that the type needs
     to implement internal features. These functions are not the same
     thing as the universal functions (ufuncs) described later. Their
     signatures can vary arbitrarily.
 
+.. c:member:: PyObject *PyArray_Descr.metadata
+
+    Metadata specific to the C implementation
+    of the particular dtype. Added for NumPy 1.7.0.
+
+.. c:member:: NpyAuxData *PyArray_Descr.c_metadata
+
+    A pointer cached with the hash value. It is -1 if not yet
+    computed. Added for NumPy 2.0.0.
```

Comment:
Hmm, I'm not sure what this field actually is. First of all, NumPy 2.0.0 doesn't exist. Second of all, as far as I can see this field is only ever set to the value -1, in other words it is never really used. That happens in `numpy/core/src/multiarray/hashdescr.c` and `numpy/core/src/multiarray/descriptor.c`.

Unless someone knows that this is actually used somewhere, probably we should just say "Currently unused. Reserved for future use in cashing hash values.".

Reply:
Also, that description should be for `hash`, not for `c_metadata`.  I think you need to shift the definitions for `metadata` and `c_metadata` down by one field, and add the `hash` field.